STYLEGUIDE.md

```diff
@@ -3,25 +3,695 @@
+
+### Declare one variable per line, wherever it makes the most sense
+
+This makes it easier to re-order the lines. However, declari variables
+wherever it makes sense (not necessairly on the top of a function).
+
+```js
+// good
+const keys = ['foo', 'bar'];
+const values = [23, 42];
+
+// bad
+const keys = ['foo', 'bar'],
+      values = [23, 42];
+```
+
+### Prefix private class methods with an underscore `JS only`
+
+Identifying private class methods makes it easier to differentiate a class's public and internal
+APIs, and makes private methods easier to mark as `private` when the code is migrated to TypeScript.
+
+```js
+// good
+class BankAccount {
+  addFunds() {}
+
+  _calculateInterest() {}
+}
+```
+
+If using TypeScript you should use the `private` modifier instead.
+
+### Magic numbers/strings
+
+These are numbers (or other values) simply used in line in your code. *Do not
+use these*, give them a variable name so they can be understood and changed
+easily.
+
+```js
+// good
+const minWidth = 300;
+
+if (width < minWidth) {
+  ...
+}
+
+// bad
+if (width < 300) {
+  ...
+}
+```
+
+### Modules
+
+Module dependencies should be written using native ES2015 syntax wherever
+possible (which is almost everywhere):
+
+```js
+// good
+import { mapValues } from 'lodash';
+export mapValues;
+
+// bad
+const _ = require('lodash');
+module.exports = _.mapValues;
+
+// worse
+define(['lodash'], function (_) {
+  ...
+});
+```
+
+In those extremely rare cases where you're writing server-side JavaScript in a
+file that does not pass run through webpack, then use CommonJS modules.
+
+In those even rarer cases where you're writing client-side code that does not
+run through webpack, then do not use a module loader at all.
+
+#### Import only top-level modules
+
+The files inside a module are implementation details of that module. They
+should never be imported directly. Instead, you must only import the top-level
+API that's exported by the module itself.
+
+Without a clear mechanism in place in JS to encapsulate protected code, we make
+a broad assumption that anything beyond the root of a module is an
+implementation detail of that module.
+
+On the other hand, a module should be able to import parent and sibling
+modules.
+
+```js
+// good
+import foo from 'foo';
+import child from './child';
+import parent from '../';
+import ancestor from '../../../';
+import sibling from '../foo';
+
+// bad
+import inFoo from 'foo/child';
+import inSibling from '../foo/child';
+```
+
+### Global definitions
+
+Don't do this. Everything should be wrapped in a module that can be depended on
+by other modules. Even things as simple as a single value should be a module.
+
+### Function definitions
+
+Use function declarations over function expressions, so that their names will
+show up in stack traces, making errors easier to debug.
+
+Also, keep function definitions above other code instead of relying on function
+hoisting.
+
+```js
+// good
+function myFunc() {
```

Comment:
interesting, lol, I always thought we preferred arrow fns.  til.

Reply:
Yeah me too and it's one of the rules I would like to put up for discussion in another or after this one is merged. Just thought if we discuss every rule AND move a lot of rules to eslint in one PR this PR would end in total chaos. So I tried to keep this PR to merging everything into one style guide and remove the rules less prime to discussion and already covered by eslint. But I'll open a couple of more PR after that around some parts of the style guide potentially being more controversial :-)